Add port configuration to Cast

Censored3 opened the pull request:
## Description:
Allow to specify port manually for media_player.cast when not using discovery:

This allows for Chromecast groups to be added when running HA in another L3 subnet.
Caveat: groups will have to tracked manually.

**Related issue (if applicable):** fixes #16124 

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#6070

## Example entry for `configuration.yaml` (if applicable):
```yaml
cast:
  media_player:
  - host: 192.168.1.10
    port: 42625
    name: My Chromecast Audio Group
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [x] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

Censored3 committed:
platform.cast: Add port configuration option - Fixes #16124

Censored3 commented:
Documentation PR: home-assistant/home-assistant.io#6070

ludeeus commented:
@Censored3 You should update the fist post with that link :) 
I makes the review prosess easier for the reviewer when all info are in the expected place :)
```
Pull request in home-assistant.github.io with documentation (if applicable): home-assistant/home-assistant.github.io#<home-assistant.github.io PR number goes here>
```

Censored3 commented:
@ludeeus Completely looked over that... Thx for the hint, I've updated the OP

balloob commented on homeassistant/components/media_player/cast.py:
```diff
@@ -224,8 +225,12 @@ def _async_create_cast_device(hass: HomeAssistantType,
+        if CONF_PORT in config:
+            ccport = config[CONF_PORT]
+        else:
+            ccport = DEFAULT_PORT
```
Add the default in the voluptuous schema, like `CONF_IGNORE_CEC`

Censored3 replied:
So, after a lot of catching up on the dev docs and guidelines, I agree this was indeed a too quick & dirty fix (after all, I just customized it for my personal needs and thought I'd just PR it as well).

Adding the default to the schema does not work. As far as I have understood, it is because of the following:
https://github.com/home-assistant/home-assistant/blob/f929c38e980d728bc74055043310bdc4c24211a1/homeassistant/components/media_player/cast.py#L201-L209

This is set by the cast platform, which did not pass through the schema:
https://github.com/home-assistant/home-assistant/blob/2f2bcf0058e9ce2e56cc06e8cf6786b44fa6c866/homeassistant/components/cast/__init__.py#L10-L14

Because of this, config[CONF_PORT] doesn't exist at all, instead of being supplied with its default value. I think this is a side-effect from when cast was moved to its own platform instead of being solely a media_player component?

An easy, but, granted, dirty fix would be to fill the CONF_PORT key in the wrapping function async_setup_entry(), or in the platform setup. I see lot of components are doing this, but it would be basically the same as my initial PR, so, not sure how to proceed here.

Censored3 commented:
Looked into it further, and realized the schema I was looking at, was in fact the schema for the deprecated style config. I added the default port there as well.

In fact, the new style config had no schema at all; I added one with a default for the port key. Also I moved the constants to a separate file.

MartinHjelmare commented on homeassistant/components/media_player/cast.py:
```diff
@@ -57,8 +57,11 @@
+    vol.Optional(CONF_PORT): cv.port,
     vol.Optional(CONF_IGNORE_CEC, default=[]):
         vol.All(cv.ensure_list, [cv.string]),
+    vol.Optional(CONF_IGNORE_CEC, default=[]): vol.All(cv.ensure_list,
```
Duplicate.

Censored3 replied:
Good catch. Wrongly fixed merge conflict.

MartinHjelmare commented on tests/components/media_player/test_cast.py:
```diff
@@ -392,8 +394,11 @@ def test_stop_discovery_called_on_stop(hass):
             return_value=mock_coro()) as mock_setup:
         await cast.async_setup_entry(hass, MockConfigEntry(), None)
 
+    print(mock_setup.mock_calls[0])
```
Stale debug print.

Censored3 replied:
That line isn't mine actually... Removed it

Censored3 commented:
Rebased on latest dev & incorporated reviews.

MartinHjelmare commented:
Tests are failing.

Censored3 commented:
> Tests are failing.

Yep, I just resubmitted, but could you please verify my change in testing is appropriate? I'm not entirely sure.
tests/components/media_player/test_cast.py @ 445

MartinHjelmare commented:
Please don't squash commits after review has started. Squashing makes it hard to follow the changes. We will squash commits when we merge.

MartinHjelmare commented on homeassistant/components/media_player/cast.py:
```diff
@@ -57,8 +57,9 @@
 
 PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
     vol.Optional(CONF_HOST): cv.string,
+    vol.Optional(CONF_PORT): cv.port,
```
Add default here too.

Censored3 committed:
Cast: testclass: fix device setups && add default port to schema of old platform, fix DEFAULT_PORT detection in old platform

Censored3 committed:
component.cast: pydocstyle recommendation

Censored3 commented:
I'll keep that in mind about not squashing commits :-)

So, everything should be okay now.

Initially, I didn't want to touch the deprecated platform media_player.cast, because, well, it's deprecated and it seems to be badly written. Don't remember how I ended up touching it anyway tbh.

Adding the default port to the schema in the platform isn't enough to make it work, I guess because it's a PLATFORM_SCHEMA and not a CONFIG_SCHEMA, which brings me back to the initial remark by @balloob, requiring a semi-hardcoded reference to the default port.
I added it again anyway, because well, the platform is deprecated anyway, and it's done the same way for the ignore_cec option. I guess the most important thing is that the cast component is completely up to standards.

MartinHjelmare commented:
The problem is that the tests are not written correctly. We should always set up the media_player component and pass the platform config. Change this in `async_cast_setup`.

MartinHjelmare commented on homeassistant/components/media_player/cast.py:
```diff
@@ -236,7 +237,7 @@ def _async_create_cast_device(hass: HomeAssistantType,
     elif CONF_HOST in config:
         info = ChromecastInfo(host=config[CONF_HOST],
-                              port=DEFAULT_PORT)
+                              port=config.get(CONF_PORT, DEFAULT_PORT))
```
The default value will have been populated by the platform schema.

Censored3 replied:
I'm pretty sure, if I remember well, it isn't just the test but the platform itself.
Not supplying a hardcoded fallback raises a KeyError on 'port'.
This is caught by the hardcoded `DEFAULT_PORT` in the original code, and also illustrated by how `ignore_cec` is handled:

https://github.com/home-assistant/home-assistant/blob/a1f08aecc5ae12e9d5e0e5e152da18e14c7c1708/homeassistant/components/media_player/cast.py#L230

Why the default values are not filled in, I don't know.

MartinHjelmare replied:
When is the `KeyError` raised? Normal set up will validate the platform config and make sure defaults are in place.

Censored3 replied:
So I'm 99,9% sure, I have had this KeyError raised within HA in my initial tries. Can't get it to trigger anymore, however.

The unittests still fail on this KeyError. I've been looking at it for quite some time but I can't figure out what to do to solve it. I can see it has something to do with passing in the config entries "statically", but I'm not sure how to solve it; I'm too inexperienced for that...

Passing in the 'port' value statically on the affected setups ofcourse solves it, but I guess it's not really a worthy test that way.
It's not that I want to get off easily, it's just that I can't figure out how to properly solve it. If you could point me in the right direction with a concrete pointer, I'd be happy to take another crack at it; otherwise, I'm afraid I'll have to throw the towel in the ring...

MartinHjelmare replied:
This is the most simple example test I could find that uses pytest functions and sets up a component with a platform.

https://github.com/home-assistant/home-assistant/blob/dev/tests/components/switch/test_unifi.py#L215-L222

So for each test we need to set up the correct component (media_player) and pass the component config dictionary including the correct platform config.

Then we assert something, most often a state in the state machine.

The existing tests use a helper function to set up the platform `async_setup_cast`. I would first see if we can modify that function to set up the component and platform as described above. Some things in the tests will have to change since we can't mock and pass `async_add_entities` to `async_setup_component`.

Censored3 commented:
@MartinHjelmare Regarding your statement about mocking `async_add_entities`:
I've gotten a step further by patching it in the HA core in `async_setup_cast`, but I just wanted to check with you if this is 'allowed'.

```
    PLATFORM_CONFIG = {'media_player': {'platform': 'cast'}}

    async def async_setup_cast(hass, config=None, discovery_info=None):
        if not config:
            config = PLATFORM_CONFIG
        else:
            config = PLATFORM_CONFIG['media_player'].update(config)

        with patch('homeassistant.helpers.entity_platform.EntityPlatform.'
                   '_async_schedule_add_entities', new=Mock()) \
                as add_entities:
            await async_setup_component(hass, 'media_player', config)
            await hass.async_block_till_done()
```

If so I just need to handle the discovery tests...

MartinHjelmare commented:
I don't think we should mock parts of the core. It's better to let the core add the entity, and then check the state of the entity.

balloob commented:
I think that we should close this PR. It's adding even more permutations of how it is configured on top of a codebase that really needs cleanup just to solve a rare edge case created by a non-supported configuration by Cast.
